src/TestQuaternion.java

```diff
@@ -1,49 +1,183 @@
+        assertEquals(new Quaternion(1.25, 1e10, 1.234e-1, -1).subtract(
+                new Quaternion(-1.25, -2e9, 1.234e-1, 1)), new Quaternion(2.5, 1.2e10, 0, -2));
+    }
+
+    /**
+     * Tester for the multiply(double) method.
+     */
+    public void testMultiplicationDouble() {
+
+        // Simple tests:
+        assertEquals(new Quaternion(1, 2, 3, 4).multiply(2),
+                new Quaternion(2, 4, 6, 8));
+        assertEquals(new Quaternion(1, 2, 3, 4).multiply(1),
+                new Quaternion(1, 2, 3, 4)); // Multiplying by identity element.
+        assertEquals(new Quaternion(1, 1, 1, 1).multiply(4.0),
+                new Quaternion(4, 4, 4, 4));
+        assertEquals(new Quaternion(1, 1, 1, 1).multiply(0.0),
+                new Quaternion(0, 0, 0, 0)); // Multiplying by zero scalar.
+        assertEquals(new Quaternion(0, 0, 0, 0).multiply(1.0),
+                new Quaternion(0, 0, 0, 0)); // Multiplying by zero quaternion.
+        assertEquals(new Quaternion(2, 2, 2, 2).multiply(0.5),
+                new Quaternion(1, 1, 1, 1)); // Multiplying by inverse of each element.
+
+
+        // Negative number tests:
+        assertEquals(new Quaternion(-1, -1, -1, -1).multiply(1),
+                new Quaternion(-1, -1, -1, -1));
+        assertEquals(new Quaternion(1, 1, 1, 1).multiply(-1),
+                new Quaternion(-1, -1, -1, -1));
+        assertEquals(new Quaternion(-1, -1, -1, -1).multiply(-1),
+                new Quaternion(1, 1, 1, 1));
+
+
+        // Mixed numbers tests:
+        assertEquals(new Quaternion(-1, 2, 3, -4).multiply(2),
+                new Quaternion(-2, 4, 6, -8));
+        assertEquals(new Quaternion(-1, 4, 70, -4).multiply(-3),
+                new Quaternion(3, -12, -210, 12));
+
+        // Floating point tests:
+        assertEquals(new Quaternion(1e1, 1e10, 1.234e-1, -1).multiply(1e10),
+                new Quaternion(1e11, 1e20, 1.234e9, -1e10));
+        assertEquals(new Quaternion(1e1, 1e10, 1.234e-1, -1).multiply(-1e10),
+                new Quaternion(-1e11, -1e20, -1.234e9, 1e10));
+    }
+
+    /**
+     * Tester for the multiply(Quaternion) method.
+     */
+    public void testMultiplicationQuaternion() {
+
+        // Simple tests:
+        assertEquals(new Quaternion(1, 2, 3, 4).multiply(new Quaternion(2, 2, 2, 2)),
+                new Quaternion(-16, 4, 12, 8));
+        assertEquals(new Quaternion(1, 2, 3, 4).multiply(new Quaternion(1, 2, 3, 4)),
+                new Quaternion(-28, 4, 6, 8));
+        assertEquals(new Quaternion(1, 1, 1, 1).multiply(new Quaternion(0, 0, 0, 0)),
+                new Quaternion(0, 0, 0, 0)); // Multiplying by right zero element.
+        assertEquals(new Quaternion(0, 0, 0, 0).multiply(new Quaternion(1, 1, 1, 1)),
+                new Quaternion(0, 0, 0, 0)); // Multiplying by left zero element.
+        assertEqualsQuaternion(new Quaternion(1, 2, 3, 4).multiply(new Quaternion(
```

Comment:
you may want to use tempoaran variables in these cases, as the test is a bit hard to read